pkg/workload/schemachange/operation_generator.go

```diff
@@ -3921,7 +3921,7 @@ func (og *operationGenerator) createFunction(ctx context.Context, tx pgx.Tx) (*o
 			return PickOne(og.params.rng, droppingEnums)
 		},
 		"ParamRefs": func() (string, error) {
-			refs, err := PickAtLeast(og.params.rng, 1, possibleParamReferences)
+			refs, err := PickBetween(og.params.rng, 1, 20, possibleParamReferences)
```

Comment:
> this is not allowed and not a particularly useful edge case to test

While I agree that testing > 100 params is not particularly interesting, what I think _is_ interesting is testing functions with exactly the maximum number of parameters, or close to that amount.

In terms of state space exploration, I think picking `{1, 20}` is not optimal. We're 20x-ing our state space but arguably going from a function with 2 params to a function with 3 params, as an example, is likely to not uncover different paths or behaviour in the DB.

IMO, a more efficient way to go would be to `PickOne` out of a few "classes" of functions:

* lowest possible number of params (1? 0?)
* few parameters (say, 3)
* "medium" amount of parameters (say 10)
* large amount of parameters (say 60)
* max allowed number of parameters (100?)

We now only have 5 paths that better cover our ability to handle different numbers of parameters in crdb.

Related question: is there a reason we don't currently test functions with 0 params?

Reply:
hihi thanks for taking a look 🙇  - I think you make a good point, but I am hesitant to say if any of those cases would really vary in behavior (in terms of code coverage). However, if anything, we could bump up 20 to be 99 and rng could cover those "classes" to a reasonable extent

Rafi might also chime in if he would like; it's very possible I'm overlooking something here

Also, we do test functions with 0 params: the "ParamRefs" key in our `placeholderMap` funcmap randomizes between using params vs not by `useParamRefs` https://github.com/cockroachdb/cockroach/blob/1898adc059e9b3026fd5a7ed88dc41cbb006abfc/pkg/workload/schemachange/operation_generator.go#L3878

used in this way

https://github.com/cockroachdb/cockroach/blob/1898adc059e9b3026fd5a7ed88dc41cbb006abfc/pkg/workload/schemachange/operation_generator.go#L3925-L3928

Reply:
> I am hesitant to say if any of those cases would really vary in behavior (in terms of code coverage)

I can understand that we might not be necessarily increasing source code coverage by much, but it does increase coverage of state space in the DB and that's what really matters. Testing functions with higher number of params would also make sure that they are handled properly outside of schema land too -- things like backup/restores, upgrade migrations, etc.

Using `PickBetween(min, max)` could be something we do as well, although I still prefer the more limited choice set, as that would test edge cases more often.

Thanks for explaining how the "no params" part of the workload works!